main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -156,29 +159,64 @@ object TypeReduction {
     val candidateMethods = clazz.getMethods
       .filter(m => m.getName == method)
       .filter(m => m.getParameterCount == ts.length)
-      .filter(m =>
-        (m.getParameterTypes zip ts).foldLeft(true){ case (acc, (clazz, tpe)) => acc && (Type.getFlixType(clazz) == tpe) } // Parameter types correspondance
-      )
+      .filter(m => // Parameter types correspondance with subtyping
+        (m.getParameterTypes zip ts).foldLeft(true) {
+          case (acc, (clazz, tpe)) => acc && isSubtype(Type.getFlixType(clazz), tpe)
+          })
+      // NB: once methods with same signatures have been filtered out, we should remove super-methods duplicates if superclass is abstract or ignore if it is a primitive type, e.g., void
+      .filter(m => m.getReturnType.isPrimitive || !java.lang.reflect.Modifier.isAbstract(m.getReturnType.getModifiers)) // temporary to avoid superclass abstract duplicate
 
     candidateMethods.length match {
+      case 0 => JavaResolutionResult.MethodNotFound
       case 1 =>
         val tpe = Type.Cst(TypeConstructor.JvmMethod(candidateMethods.head), loc)
         JavaResolutionResult.Resolved(tpe)
-      case _ => JavaResolutionResult.MethodNotFound // For now, method not found either if there is an ambiguity or no method found
+      case _ =>
+        // Among candidate methods if there already exists the method with the exact signature, we should ignore the rest. E.g.: append(String), append(Object) in SB for the call append("a") should already know to use append(String)
+        // TODO: there is surely a better way to repeat the operation, pre-factorization should be possible
+        val exactSignature = candidateMethods
+          .filter(m => // Parameter types correspondance with subtyping
+            (m.getParameterTypes zip ts).foldLeft(true) {
+              case (acc, (clazz, tpe)) => acc && (Type.getFlixType(clazz) == tpe)
+            })
+        exactSignature.length match {
+          
+          
+          case 1 => JavaResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmMethod(exactSignature.head), loc))
+          case _ => JavaResolutionResult.AmbiguousMethod(candidateMethods.toList)
```

Comment:
What happens if it is 0?

Reply:
Inside the ambiguous case we have at least two candidate methods. If there is one among them that has the exact signature then we pick it. Otherwise there remain the candidate methods that cause ambiguity => ambiguous error.
In particular this case is indeed binary: 0 then it means the user has to do a proper uppercast to select the method he wants, 1 then the exact method exists, 2 or more cannot happen as it would be already prohibited by Java since there would be a fixed ambiguity.